python/paddle/fluid/layers/collective.py

```diff
@@ -178,3 +183,223 @@ def _c_sync_comm_stream(x, ring_id):
         outputs={'Out': [x]},
         attrs={'ring_id': ring_id})
     return x
+
+
+class DistributedClassifier(object):
+    '''
+    Tookit for distributed classification, in which the parameter of the last
+    full-connected layer is distributed to all trainers
+    '''
+
+    def __init__(self, nclasses, nranks, rank_id):
+        self.nclasses = nclasses
+        self.nranks = nranks
+        self.rank_id = rank_id
+
+        self.shard_dim = (nclasses + nranks - 1) // nranks
+        self.padding_dim = 0
+        self.is_equal_division = True
+        if nclasses % nranks != 0:
+            self.is_equal_division = False
+            if rank_id == nranks - 1:
+                other_shard_dim = self.shard_dim
+                self.shard_dim = nclasses % other_shard_dim
+                self.padding_dim = other_shard_dim - self.shard_dim
+
+    def create_parameter(self,
+                         in_dim,
+                         param_attr=None,
+                         transpose_weight=False,
+                         use_bias=True):
+        helper = LayerHelper("dist_fc", **locals())
+        dtype = helper.input_dtype()
+
+        if param_attr is None:
+            stdv = math.sqrt(2.0 / (in_dim + out_dim))
+            param_attr = ParamAttr(initializer=Normal(scale=stdv))
+        weight_shape = [self.shard_dim, in_dim
+                        ] if transpose_weight else [in_dim, self.shard_dim]
+        weight = helper.create_parameter(
+            shape=weight_shape, dtype=dtype, attr=param_attr, is_bias=False)
+        # avoid distributed parameter allreduce gradients
+        weight.is_distributed = True
+        # avoid distributed parameter broadcasting in startup program
+        default_startup_program().global_block().vars[
+            weight.name].is_distributed = True
+
+        bias = None
+        if use_bias:
+            bias = helper.create_parameter(
+                shape=shard_dim, dtype=dtype, is_bias=True)
```

Comment:
Should shape be in_dim if transpose_weight is True?

Reply:
not necessary